Assets/Scripts/InventoryManager.cs

```diff
@@ -0,0 +1,135 @@
+
+public class InventoryManager : MonoBehaviour
+{
+    public static InventoryManager Instance;
+
+    //This will be the list that things are added and removed to via purchasing items
+    [SerializeField] private List<PlaceableObject> inventoryPlaceableObjects = new List<PlaceableObject>();
+    [SerializeField] private GameObject inventoryGUI;
+    [SerializeField] private GameObject inventoryGrid;
+    [SerializeField] private GameObject inventoryItemPrefab;
+    [SerializeField] private GameObject playerHeldItemParent;
+    [SerializeField] private float stockScale;
+    [SerializeField] private float structureScale;
+
+
+    private PlaceableObject heldObject;
+    public PlaceableObject HeldObject { get => heldObject; set => heldObject = value; }
+    private int tabIndex;
+    private List<GameObject> gridObjectDisplayList = new List<GameObject>();
+    //List of items currently displayed in GUI
+    private List<PlaceableObject> inventoryObjectDisplayList = new List<PlaceableObject>();
+    private GameObject playerHeldItem;
+    public GameObject PlayerHeldItem { get => playerHeldItem; set => playerHeldItem = value; }
+
+    // Makes InventorManager a singleton
+    private void Awake()
+    {
+        if (Instance != null)
+        {
+            Destroy(this);
+            return;
+        }
+
+        Instance = this;
+    }
+
+    // Sets inventory to be closed when the game starts
+    void Start()
+    {
+        inventoryGUI.SetActive(false);
+        HeldObject = null;
+        InputSystem.actions.FindAction("ToggleInventory").performed += ctx => ToggleInventoryGUI();
+        SwitchTab(0);
+    }
+    
+    // This toggles the state of the Inventory GUI (open or closed)
+    private void ToggleInventoryGUI()
+    {
+        inventoryGUI.SetActive(!inventoryGUI.activeSelf);
+
+        if (inventoryGUI.activeSelf)
+        {
+            Cursor.lockState = CursorLockMode.None;
+            SwitchTab(tabIndex);
+            ClearHandItem();
+        }
+        else
+        {
+            Cursor.lockState = CursorLockMode.Locked;
+        }
+    }
+
+    // Changes the tab and resets the contents of the inventory
+    public void SwitchTab(int index)
+    {
+        tabIndex = index;
+        SetInventoryDisplayContent();
+    }
+
+    // Sets all of the content of the inventory GUI
+    private void SetInventoryDisplayContent()
+    {
+        // Removes all the old GUI display gridItems, and clears the lists of what was in them
+        gridObjectDisplayList.ForEach(gridItem => Destroy(gridItem));
+        gridObjectDisplayList.Clear();
+        inventoryObjectDisplayList.Clear();
+
+        int indexCount = 0;
+
+        // Iterates over all stock to see if it should display in current tab
+        inventoryPlaceableObjects.ForEach(placeableObject => {
+            if (((int)placeableObject.type) == tabIndex)
+            {
+                CreateGridItem(indexCount, placeableObject);
+
+                indexCount++;
```

Comment:
would this be better suited to a normal for loop, that uses `inventoryPlacableObjects.Count`,
so you don't have to make the `indexCount` variable and force it to iterate?

Or you could use a quirky foreach I found on [stackoverflow](https://stackoverflow.com/a/11437562), which implements `i` into the foreach statement

Reply:
In this case that wouldn't quite work. The indexCount only increases if the criteria is met as the number is used assuming that is the case so a normal for loop or the special foreach would not work as intended.

Reply:
Ah my bad, i misread the code